.crawler/src/crawler.rs

```diff
@@ -0,0 +1,274 @@
+            }
+            ClientMessage::PeerRequest => self.process_peer_request(source).await?,
+            ClientMessage::PeerResponse(peer_ips) => self.process_peer_response(source, peer_ips).await?,
+            ClientMessage::Ping(version, _fork_depth, _peer_type, _peer_state, _block_hash, block_header) => {
+                // Deserialise the block header.
+                let block_header = block_header.deserialize().await.unwrap();
+                self.process_ping(source, version, block_header.height()).await?
+            }
+            ClientMessage::Pong(_is_fork, _block_locators) => {}
+            ClientMessage::UnconfirmedBlock(_block_height, _block_hash, _block) => {}
+            ClientMessage::UnconfirmedTransaction(_transaction) => {}
+            ClientMessage::PoolRegister(_address) => {}
+            ClientMessage::PoolRequest(_share_difficulty, _block_template) => {}
+            ClientMessage::PoolResponse(_address, _nonce, _proof) => {}
+            _ => return Err(io::ErrorKind::InvalidData.into()), // Peer is not following the protocol.
+        }
+
+        Ok(())
+    }
+}
+
+// Helper methods.
+impl Crawler {
+    async fn process_peer_request(&self, source: SocketAddr) -> io::Result<()> {
+        let peers = vec![];
+        let msg = ClientMessage::PeerResponse(peers);
+        info!(parent: self.node().span(), "sending a PeerResponse to {}", source);
+
+        self.send_direct_message(source, msg)?;
+
+        Ok(())
+    }
+
+    async fn process_peer_response(&self, source: SocketAddr, mut peer_ips: Vec<SocketAddr>) -> io::Result<()> {
+        let node = self.clone();
+        task::spawn(async move {
+            peer_ips.retain(|addr| node.node().listening_addr().unwrap() != *addr);
+
+            // Insert the address into the known network and update the crawl state.
+            node.known_network.update_connections(source, peer_ips.clone());
+            node.known_network.received_peers(source);
+
+            for peer_ip in peer_ips {
+                if !node.node().is_connected(peer_ip) && !node.state.peers.lock().await.iter().any(|peer| peer.listening_addr == peer_ip) {
+                    info!(parent: node.node().span(), "trying to connect to {}'s peer {}", source, peer_ip);
+
+                    // Only connect if this address is unknown.
+                    if !node.known_network.nodes().contains_key(&peer_ip) {
+                        let _ = node.node().connect(peer_ip).await;
+                    }
+                }
+            }
+        });
+
+        Ok(())
+    }
+
+    async fn process_ping(&self, source: SocketAddr, version: u32, block_height: u32) -> io::Result<()> {
+        // Ensure the message protocol version is not outdated.
+        if version < MESSAGE_VERSION {
+            warn!(parent: self.node().span(), "dropping {} due to outdated version ({})", source, version);
+            return Err(io::ErrorKind::InvalidData.into());
+        }
```

Comment:
TODO: we should probably maintain a detailed list of non-compliant peers so we can report their numbers and reasons for non-compliance with the protocol